presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+    private int keyCount;       // number of valid entries in keyIndices array
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<SelectiveStreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+
+    private Block keyBlock;
+    private int readOffset;
+    private int[] nestedReadOffsets;
+
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+    private int[] offsets;
+    private boolean[] nulls;
+    private int[] outputPositions;
+    private int outputPositionCount;
+    private boolean outputPositionsReadOnly;
+    private boolean allNulls;
+    private boolean valuesInUse;
+
+    private int[] nestedLengths;
+    private int[][] nestedPositions;
+    private int[] nestedPositionCounts;
+    private int[][] nestedOutputPositions;
+    private boolean[][] inMap;
+
+    private final AggregatedMemoryContext systemMemoryContext;
+    private final LocalMemoryContext localMemoryContext;
+
+    public MapFlatSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Map<Subfield, TupleDomainFilter> filters,
+            List<Subfield> requiredSubfields,
+            Optional<Type> outputType,
+            DateTimeZone hiveStorageTimeZone,
+            AggregatedMemoryContext systemMemoryContext)
+    {
+        checkArgument(filters.keySet().stream().map(Subfield::getPath).allMatch(List::isEmpty), "filters on nested columns are not supported yet");
+
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getOrcTypeKind();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.localMemoryContext = systemMemoryContext.newLocalMemoryContext(MapFlatSelectiveStreamReader.class.getSimpleName());
+        this.outputRequired = requireNonNull(outputType, "outputType is null").isPresent();
+        this.outputType = outputType.map(MapType.class::cast).orElse(null);
+
+        TupleDomainFilter filter = getTopLevelFilter(filters).orElse(null);
+        this.nullsAllowed = filter == null || filter.testNull();
+        this.nonNullsAllowed = filter == null || filter.testNonNull();
+
+        requireNonNull(requiredSubfields, "requiredSubfields is null");
+        switch (keyOrcType) {
```

Comment:
what about the other orckind DATE, Timestamp, float, double ...

Reply:
@bhhari Flat map encoding only supports int and string keys. Date, floats etc. keys are not supported.

Reply:
@bhhari Flat map encoding only supports int and string keys. Date, floats etc. keys are not supported.